models/exposure.py

```diff
@@ -1,7 +1,238 @@
+
+    width = sa.Column(sa.Integer, nullable=False, index=False, doc="Width of each CCD in pixels. ")
+
+    height = sa.Column(sa.Integer, nullable=False, index=False, doc="Height of each CCD in pixels. ")
+
+    project = sa.Column(sa.Text, index=True, nullable=False, doc='Name of the project, (could also be a proposal ID). ')
+
+    target = sa.Column(sa.Text, index=True, nullable=False, doc='Name of the target object or field id. ')
+
+    gallat = sa.Column(sa.Double, index=True, doc="Galactic latitude of the target. ")
+
+    gallon = sa.Column(sa.Double, index=False, doc="Galactic longitude of the target. ")
+
+    ecllat = sa.Column(sa.Double, index=True, doc="Ecliptic latitude of the target. ")
+
+    ecllon = sa.Column(sa.Double, index=False, doc="Ecliptic longitude of the target. ")
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize the exposure object.
+        Can give the filename of the exposure
+        as the single positional argument.
+
+        Otherwise, give any arguments that are
+        columns of the Exposure table.
+
+        If the filename is given, it will parse
+        the instrument name from the filename.
+        The header will be read out from the FITS file.
+        """
+        super().__init__(**kwargs)  # put keywords into columns
+
+        self._data = None
+        self.instrument_object = None
+
+        if len(args) == 1 and isinstance(args[0], str):
+            self.filename = args[0]
+
+        if self.instrument is None:
+            self.parse_instrument_name()
+
+        # find additional column values from the header
+        self.read_header()
+
+        self.num_ccds = self.get_instrument_object().get_num_ccds()
+
+        # override the header values with any given keyword arguments?
+        for key, value in kwargs.items():
+            if key in self.__dict__:
+                setattr(self, key, value)
+
+        if self.ra is not None and self.dec is not None:
+            self.calculate_coordinates()
+
+    @orm.reconstructor
+    def init_on_load(self):
+        self._data = None
+        self.instrument_object = None
+
+    def parse_instrument_name(self):
+        if self.filename is None:
+            return
+
+        for name, class_ in inspect.getmembers(models.instruments, inspect.isclass):
+            if name.lower() in self.filename.lower():
+                self.instrument = name
+                self.instrument_object = class_()
+                break
+
+    def __repr__(self):
+        return f"Exposure(id: {self.id}, exp: {self.exp_time}s, filt: {self.filter}, from: {self.instrument}/{self.telescope})"
+
+    def __str__(self):
+        return self.__repr__()
+
+    def save(self):
+        pass  # TODO: implement this! do we need this?
+
+    def load(self, ccd_ids=None):
+        if ccd_ids is None:
+            ccd_ids = list(range(self.num_ccds))
```

Comment:
While we can choose to define our internal ccd_ids to be indexed starting from 0, it might be convenient if we made that indexing an instrument-specific thing.  (That way, we don't have to maintain our own map of ccd id to ccd identifier.)  Different instruments are going to have different ways of specifying their CCDs.  E.g., ztf has a ccd indentifier that's like a coordinate position, rather than just an index.  Perhaps instead of list(range(self.num_ccds)), we could cal an instrument's all_ccds() method or some such.

Another thing: it might be worth having a ccd table, with a foreign key that points back to the instrument table.  This could store things like offsets of the corners (and center) of the CCD relative to the overall exposure coordinates.  (This could also just be encoded into the class, not in a table.  One advantage of having it in a table instead of just in code is that you might be able to do database searches on those parameters.)